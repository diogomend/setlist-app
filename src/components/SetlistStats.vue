<template>
  <div class="setlist-stats">
    <v-layout wrap="wrap">
      <v-flex xs12="xs12">
        <v-layout wrap="wrap">
          <v-flex xs6 md3>
            <v-card class="text-xs-center" height="100%">
              <v-card-text>
                <div class="display-1 mb-2">{{ totalConcerts }}</div>
                Concerts
              </v-card-text>
            </v-card>
          </v-flex>
          <v-flex xs6 md3>
            <v-card class="text-xs-center" height="100%">
              <v-card-text>
                <div class="display-1 mb-2">{{ totalArtists }}</div>
                Artists
              </v-card-text>
            </v-card>
          </v-flex>
          <v-flex xs6 md3>
            <v-card class="text-xs-center" height="100%">
              <v-card-text>
                <div class="display-1 mb-2">{{ totalCountries }}</div>
                Countries
              </v-card-text>
            </v-card>
          </v-flex>
          <v-flex xs6 md3>
            <v-card class="text-xs-center" height="100%">
              <v-card-text>
                <div class="display-1 mb-2">{{ currentYear }}</div>
                This year
              </v-card-text>
            </v-card>
          </v-flex>
        </v-layout>
      </v-flex>
    </v-layout>
  </div>
</template>

<script>
import { mapGetters } from "vuex";
export default {
  name: "SetlistStats",
  computed: {
    ...mapGetters({
      yearlyConcerts: "setlist/getYearlyConcerts",
      totalConcerts: "setlist/getTotalConcerts",
      totalArtists: "setlist/getTotalArtists",
      totalCountries: "setlist/getTotalCountries"
    }),
    currentYear() {
      const year = new Date().getFullYear();
      console.log(this.yearlyConcerts);
      return typeof this.yearlyConcerts[year] !== "undefined"
        ? this.yearlyConcerts[year].length
        : 0;
    }
  }
};
</script>
