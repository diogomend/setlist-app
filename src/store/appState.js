const actions = {
  toggleDrawer: ({ commit }, toggle) => {
    commit("TOGGLE_DRAWER", toggle);
  },
  toggleLoading: ({ commit }, toggle) => {
    commit("TOGGLE_LOADING", toggle);
  }
};

const mutations = {
  TOGGLE_DRAWER(state, bool) {
    state.drawer = bool;
  },
  TOGGLE_LOADING(state, isLoading) {
    state.isLoading = isLoading;
  }
};

const getters = {
  getDrawer: state => state.drawer,
  getLoading: state => state.isLoading
};

const state = () => ({
  drawer: false,
  isLoading: true
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
