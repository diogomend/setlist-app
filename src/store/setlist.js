import setlistService from "@/service/setlist";

const actions = {
  fetchSetlists: async ({ commit, dispatch }) => {
    try {
      const { data } = await setlistService.getSetlists();
      commit("UPDATE_SETLISTS", data);
      commit("UPDATE_ARTISTS", data);
      //return Promise.resolve(data);
      return true;
    } catch (err) {
      //return Promise.reject('Could not load setlist information');
      return false;
    } finally {
      dispatch("appState/toggleLoading", false, { root: true })
    }
  }
};

const mutations = {
  UPDATE_SETLISTS(state, data) {
    state.setlists = data;
  },
  UPDATE_ARTISTS(state, data) {
    state.artists = getArtists(data);
  }
};

const getters = {
  getTotalConcerts: state => state.setlists.length,
  getShow: state => id => {
    return Object.values(state.setlists).find(element => element.id === id);
  },
  getArtist: state => mbid => {
    const data = Object.values(state.setlists).find(
      element => element.artist.mbid === mbid
    );
    if (data && Object.prototype.hasOwnProperty.call(data, "artist")) {
      return data.artist;
    }

    return null;
  },
  getArtists: state => {
    const artists = getArtists(state.setlists);

    return Object.values(artists).sort((a, b) => (a.name > b.name ? 1 : -1));
  },
  getArtistShows: state => mbid => {
    return state.setlists.reduce((ac, setlist) => {
      if (setlist.artist.mbid === mbid) {
        ac.push(setlist);
      }

      return ac;
    }, []);
  },
  getTopArtists: state => {
    const artists = Object.values(state.artists).sort((a, b) =>
      a.qty < b.qty ? 1 : -1
    );
    return artists.slice(0, 6);
  },
  getTotalArtists: state => Object.keys(state.artists).length,
  getYearlyConcerts: state => {
    return state.setlists.reduce((ac, setlist) => {
      const dateSplit = setlist.eventDate.split("-");
      const year = parseInt(dateSplit[2]);

      if (Object.prototype.hasOwnProperty.call(ac, year)) {
        const merge = [...ac[year], setlist];
        return { ...ac, [year]: merge };
      }

      return { ...ac, [year]: [setlist] };
    }, {});
  },
  getTotalCountries: state => {
    const countries = state.setlists.reduce((ac, setlist) => {
      const country = setlist.venue.city.country.code;
      const qty = Object.prototype.hasOwnProperty.call(ac, country)
        ? ac[country] + 1
        : 1;

      return { ...ac, [country]: qty };
    }, {});

    return Object.keys(countries).length;
  }
};

const state = () => ({
  setlists: [],
  isLoading: true
});

const getArtists = setlists => {
  return setlists.reduce((ac, setlist) => {
    setlist.artist.qty = Object.prototype.hasOwnProperty.call(
      ac,
      setlist.artist.mbid
    )
      ? ac[setlist.artist.mbid].qty + 1
      : 1;

    return { ...ac, [setlist.artist.mbid]: setlist.artist };
  }, {});
};

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
