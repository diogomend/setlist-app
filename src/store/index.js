import Vue from "vue";
import Vuex from "vuex";
import setlist from "@/store/setlist";
import appState from "@/store/appState";

Vue.use(Vuex);

export const storeOptions = {
  modules: {
    setlist,
    appState
  }
};

export default new Vuex.Store(storeOptions);
