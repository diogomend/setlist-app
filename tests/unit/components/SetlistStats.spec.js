import { shallowMount } from "@vue/test-utils";
import Vue from "vue";
import { storeOptions } from "@/store/index";
import SetlistStats from "@/components/SetlistStats";
import Vuex from "vuex";
import Vuetify from "vuetify";

Vue.use(Vuex);
Vue.use(Vuetify);
describe("SetlistStats.vue", () => {
  let store;
  let localVue;

  beforeEach(() => {
    jest.clearAllMocks();
    storeOptions.modules.setlist.state = {
      setlists: [
        {
          eventDate: "20-06-2016",
          venue: {
            city: { country: { code: "PT" } }
          }
        }
      ],
      artists: {}
    };
    store = new Vuex.Store(storeOptions);
  });

  it("is a Vue instance", () => {
    const wrapper = shallowMount(SetlistStats, {
      localVue,
      store
    });

    const template = wrapper.html();
    expect(template).toMatchSnapshot();
  });

  it("should return number of concerts in current year", async () => {
    const mockDate = new Date(1466424490000);
    jest.spyOn(global, "Date").mockImplementation(() => mockDate);
    const wrapper = shallowMount(SetlistStats, {
      localVue,
      store
    });

    expect(wrapper.vm.currentYear.length).toEqual(1);
  });
});
