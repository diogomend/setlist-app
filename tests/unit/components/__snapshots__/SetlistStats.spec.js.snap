// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetlistStats.vue is a Vue instance 1`] = `
<div class="setlist-stats">
  <v-layout-stub tag="div" wrap="wrap">
    <v-flex-stub tag="div" xs12="xs12">
      <v-layout-stub tag="div" wrap="wrap">
        <v-flex-stub tag="div" xs6="" md3="">
          <v-card-stub loaderheight="4" tag="div" height="100%" class="text-xs-center">
            <v-card-text-stub>
              <div class="display-1 mb-2">1</div>
              Concerts
            </v-card-text-stub>
          </v-card-stub>
        </v-flex-stub>
        <v-flex-stub tag="div" xs6="" md3="">
          <v-card-stub loaderheight="4" tag="div" height="100%" class="text-xs-center">
            <v-card-text-stub>
              <div class="display-1 mb-2">0</div>
              Artists
            </v-card-text-stub>
          </v-card-stub>
        </v-flex-stub>
        <v-flex-stub tag="div" xs6="" md3="">
          <v-card-stub loaderheight="4" tag="div" height="100%" class="text-xs-center">
            <v-card-text-stub>
              <div class="display-1 mb-2">1</div>
              Countries
            </v-card-text-stub>
          </v-card-stub>
        </v-flex-stub>
        <v-flex-stub tag="div" xs6="" md3="">
          <v-card-stub loaderheight="4" tag="div" height="100%" class="text-xs-center">
            <v-card-text-stub>
              <div class="display-1 mb-2">0</div>
              This year
            </v-card-text-stub>
          </v-card-stub>
        </v-flex-stub>
      </v-layout-stub>
    </v-flex-stub>
  </v-layout-stub>
</div>
`;
